app/models/restaurant.rb

```diff
@@ -0,0 +1,80 @@
+##
+# This class represents a Restaurant
+#
+# == Schema Information
+#
+# Table name: restaurants
+#
+#  id         :integer          not null, primary key
+#  name       :string
+#  created_at :datetime         not null
+#  updated_at :datetime         not null
+#  api_key    :uuid
+#
+
+class Restaurant < ApplicationRecord
+    has_many :meals, dependent: :destroy
+    has_many :assignments
+    has_many :delivery_zones, through: :assignments
+    has_many :restrict_restaurant_delivery_zones, dependent: :destroy
+    has_many :restrict_restaurant_days, dependent: :destroy
+    before_create :generate_api_key
+    after_destroy :destroy_future_assignments
+
+    ##
+    # Returns restaurants ordered by those with the last amount of historical delivery assignments ascending, while also
+    # filtering out specific ones based on biz rules
+    # Params:
+    # delivery_zone_id: The ID of delivery zone to sort number of assignments to by
+    # day: The day for which the assignment is going to be for
+    #   Restaurant.find_least_assigned_available(delivery_zone_id, Date.today)
+    def self.find_least_assigned_available(delivery_zone_id, day)
+        least_assigned = self
+            .find_least_assigned(delivery_zone_id)
+            .having("sum(case when assignments.date = '#{day}' THEN 1 ELSE 0 end) < 3") # Having less than 3 delivery zones in the same day
+            .having("sum(case when assignments.delivery_zone_id = #{delivery_zone_id} THEN 1 ELSE 0 end) < 1") # Having not already delivered to this delivery zone
+        least_assigned = least_assigned.where('restaurants.id not in (?)', self.restricted_ids(delivery_zone_id, day)) unless self.restricted_ids(delivery_zone_id, day).blank?
+        least_assigned.first
+    end
+
+    private 
+
+    ##
+    # Destroys only the assignments in the future
+    def destroy_future_assignments
+        self.assignments.in_the_future.destroy_all
+    end
+
+    ##
+    # Generates the unique API key for the restaurant
+    def generate_api_key
+        begin
+            self.api_key = SecureRandom.uuid
```

Comment:
Since we are storing this api key in the db, shouldn't we encrypt it? Alternatively, you could also use encrypted global id, which might be somewhat simpler to manage.

Reply:
I was actually thinking about that. I forgot to put my thoughts in the ```README``` unfortunately. If I encrypt it, performance would take a hit since I'd have to decrypt and search though the database when verifying the API key. BUT after thinking about it now, it would have been easier to encrypt the input API key and search by that. :\

Reply:
As you mentioned, no sane person would run bcrypt in a full table scan - which is the very thing bcrypt is designed to prevent against. You can, however, add a second identifier key to create a username/password kind of combination. Of course, nowadays the accepted approach is to use perishable access tokens and etc. but that's probably beyond the scope of this challenge.

Regardless, I would encourage you to look into [global id](https://github.com/rails/globalid), which was developed by the rails core team. It provides a very simple solution for something like this similar to JWT.